### PR TITLE
Manage Oracle extension pack

### PR DESCRIPTION
**Pull request by mark0n**
I added support for installing/removing Oracle's extension pack.

Note that there is a small limitation though: Although the latest VirtualBox package can be installed automatically using `package_ensure => latest`, clean and fully automatic upgrades of the extension pack are not possible right now. The reasons for this are:

- We need the full version string (version + build as reported by `VBoxManage --version`) of VirtualBox to compose the download URL. However, I did not find a way to automatically determine the full version string in `virtualbox::extpack`.
- Facter cannot be used for this purpose since it runs _before_ `virtualbox::install` installs/upgrades VirtualBox but we need to know which version is installed _after_ `virtualbox::install` is finished (`VBoxManage` might not even be available while Facter is running). If installing/upgrading the extension pack one Puppet run after installing/upgrading VirtualBox is acceptable you can cheat and set the version to `$virtualbox_version`. Note that this is ugly and might leave you without a usable extension pack for 30 minutes...

Bottom line is: no automatic upgrades so far, you have to bump up the version in your manifest file manually. If you have any ideas how to work around this let me know!

Note that I'm not a Ruby programmer and not familiar with writing tests. I would appreciate if someone could help me out here ;-)

### CODE REVIEW

**Comment by danzilio** [comment]
Hey Martin! Thanks for the patch! I like the idea, I do have some suggestions though. Let me make some notes and I'll get back to you later today.

**Comment by mark0n** [comment]
I just fixed a small issue with the new fact which caused the extension pack to be re-installed on every Puppet run. Let me know if you have any suggestions to improve the code.

**Comment by mark0n** [comment]
I just fixed a small issue with the new fact which caused the extension pack to be re-installed on every Puppet run. Let me know if you have any suggestions to improve the code.

**Comment by danzilio** [comment]
So, my initial thought is to generalize this a bit. How about we let the `virtualbox::extpack` class take a `source` parameter so users can define their own source for the extpack? In this case, the class would ignore the `version` parameter and just grab the specific file defined by the user. Also, how about we use `maestrodev/wget` instead of an exec here? That way we can be sure that `wget` is available on the system and we can define the `wget` statement natively in the DSL. What do you think?

**Comment by mark0n** [comment]
Thanks for providing some input on this. Using the `wget` class definitely makes sense. I'm not sure if I understand your comment regarding the `source` parameter comment, though. Do you want me to generalize it to make it possible to install other extension packs (I haven't heard of extension packs provided by a third party so far)? Or do you just want to give users the option to customize the download URL? Note that I still need the `version` parameter to determine if the extension pack needs to be upgraded.

**Comment by danzilio** [comment]
For me, as an enterprise user, I want to be able to define the source of the file that I'm downloading with this module (for security and convenience purposes). I think allowing the user to set `version` is great if they're not too particular and want the stock Oracle VirtualBox Extension pack, but if you want more power to define where the extension pack comes from, the `source` parameter becomes an essential feature. For example:

```ruby
class { 'virtualbox::extpack':
  version => '4.3.20',
}
```

or...

```ruby
class { 'virtualbox::extpack':
  source => 'http://myserver.example.com/Oracle_VM_VirtualBox_Extension_Pack-4.3.20.vbox-extpack',
}
```

In the first case, the class would download the file based on the `version` parameter. Upgrades would be prompted if the `version` parameter is changed. In the second case, the class would download the file based on the `source` parameter and upgrades would be prompted if the `source` parameter is changed. In the case that I specify both the `source` and `version` parameters, the `source` parameter should take precedence. Make sense?

I know there are some people out there who have created third party extension packs for VirtualBox, but I'm not aware of many people using them. If we get some feedback that people want the ability to install other extension packs, we could convert this to a defined type. But, for now, I think this is a good feature to add to the module as-is.

Thanks for working on this! I appreciate your input.

File: manifests/extpack.pp [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        # the same Puppet run)
+        if $virtualbox::package_ensure == 'absent' {
+          Exec['uninstall vbox extpack'] -> Class['virtualbox::install']
+        } else {
+          Class['virtualbox::install'] -> Exec['uninstall vbox extpack']
+        }
+      }
+    }
+    'default': {
+      fail('Invalid value for attribute \'ensure\'')
```
Comment (danzilio) [inline manifests/extpack.pp]: This can be written as: `fail("Invalid value for attribute 'ensure'")`

Double quotes are ok for interpolation or for cases where you need to use single quotes. I'm not sure if `puppet-lint` picks up on this. If not, we can tell it to ignore this line.

**Comment by mark0n** [comment]
Installing from a custom URL might make sense in some environments. However, if you only provide the `source` parameter I cannot think of a way to figure out if the URL points to something different from what is installed (we don't want the extension pack to be installed on every Puppet run).

For this reason you need to provide the `version` parameter along with `source`:
```
class { 'virtualbox::extpack':
  version => '4.3.20r96996',
  source  => 'http://myserver.example.com/Oracle_VM_VirtualBox_Extension_Pack-4.3.20-96996.vbox-extpack',
}
```
Does the new version work for you? (I also fixed some other minor issues and cleaned up the Git history.)

File: manifests/params.pp [hunk 2]
```diff
@@ -9,6 +9,9 @@
   $manage_package = true
   $manage_repo = true
   $package_ensure = present
+  $extpack_ensure = present
+  $extpack_version = $::virtualbox_version
+  $extpack_source = ''
```
Comment (danzilio) [inline manifests/params.pp]: Let's leave the `extpack_source` parameter out. This way, it'll be set to `undef` instead of an empty string.

File: manifests/extpack.pp [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+      if $virtualbox::package_ensure == 'absent' {
+        fail('Cannot install extension pack without installing VirtualBox')
+      }
+
+      # install/upgrade if we do not have the requested version
+      if $::virtualbox_extpack_version != $version {
+        $ver = regsubst($version, '^([^r]+)r([^r]+)', '\1')
+        $build = regsubst($version, '^([^r]+)r([^r]+)', '\2')
+        $filename = "Oracle_VM_VirtualBox_Extension_Pack-${ver}-${build}.vbox-extpack"
+        if $source != '' {
```
Comment (danzilio) [inline manifests/extpack.pp]: Once we remove the `source` parameter (see my comment on the params class), we can change this logic to:

```ruby
if $source {
  ...
} else {
 ...
}
```

File: manifests/extpack.pp [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+#
+# [*source*]
+#   Download extension pack from the given URL. If not specified download from
+#   Oracle's official server.
+#
+class virtualbox::extpack (
+  $ensure  = $virtualbox::params::extpack_ensure,
+  $version = $virtualbox::params::extpack_version,
+  $source  = $virtualbox::params::extpack_source
+) {
```
Comment (danzilio) [inline manifests/extpack.pp]: Also, let's make sure to inherit `params` here.

**Comment by danzilio** [comment]
I get what you're saying. I think you're right, the `version` parameter provides the most integrity for the installed extension pack. There may be a better way to do it, but that's something that can be worked out in a later release. Could you document the parameters in `extpack.pp` (take a look at the parameter documentation in `init.pp` for examples)? Also, do you want me to walk you through the rspec tests?

**Comment by mark0n** [comment]
Thanks for your comments. I added a couple of rspec tests that should cover the most important things (and immediately revealed two smaller issues that I have fixed as well). Since these are my first rspec tests I would appreciate your comments.

My class contains some code that ensures correct order of `class['virtualbox::install']` and (un)installing the extension pack. I think this should be covered by tests as well. I'll look into that tomorrow.

**Comment by danzilio** [comment]
Some notes on the tests:

`virtualbox_extpack_spec.rb` should actually be `virtualbox__extpack_spec.rb` (note the two underscores to match the namespace separator in Puppet). I also want to make sure we're testing all of the logic and interfaces for the `virtualbox::extpack` class. I see there are a few test cases missing, but they should be easy to add. I would add test cases to catch when no source parameter is passed (this particular case won't pass, more on this in a sec). I think you've identified a good base set of test cases, but we should also test for those cases when the user doesn't provide the source. See my inline comments for more.

This is shaping up to be a really good PR! Looking forward to getting it merged in!

File: manifests/extpack.pp [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+#   Oracle's official server.
+#
+class virtualbox::extpack (
+  $ensure  = $virtualbox::params::extpack_ensure,
+  $version = $virtualbox::params::extpack_version,
+  $source  = $virtualbox::params::extpack_source
+) inherits virtualbox::params {
+  validate_string($ensure)
+  validate_string($version)
+  validate_string($source)
```
Comment (danzilio) [inline manifests/extpack.pp]: This effectively makes 'source' required. If it's undef, it's not a string. We should change this to read:

```
if $source {
  validate_string($source)
}
```

**Comment by mark0n** [comment]
Done.

Is there a way to test the order between sub-classes of virtualbox? Or is this beyond the scope of rspec tests?

**Comment by danzilio** [comment]
Yes! The relationship matchers help with this. For example:

```
it { should contain_exec('install vbox extpack').that_requires('Class[virtualbox::install]') }
it { should contain_exec('install vbox extpack').that_requires('Wget::Fetch[download vbox extpack]') }
```

You can also use the `that_comes_before` method to express the converse:

```
it { should contain_class('virtualbox::install').that_comes_before('Exec[install vbox extpack]') }
it { should contain_wget__fetch('download vbox extpack').that_comes_before('Exec[install vbox extpack]') }
```

**Comment by danzilio** [comment]
One thing I'll note about the relationship matchers is that, as of right now, those will fail. None of the `virtualbox` classes are included in the catalog yet. You'll need to add `let(:pre_condition) { 'include virtualbox' }` to the context; this ensures that the `virtualbox` class is added to the catalog.

At this point, the Debian cases will begin to fail because the `apt` module uses the `lsbdistid` and `lsbdistcodename` facts, but they aren't set here. Those will need to be added as well. You can set `lsbdistid` to the same value as `osfamily` and `lsbdistcodename` just needs to be a string, you can set it to anything you'd like.

**Comment by mark0n** [comment]
Thanks for providing the `let(:pre_condition) { 'include virtualbox' }` - that was exactly what I was looking for :-) I just pushed my implementation. Let me know if it needs further clean up.

**Comment by mark0n** [comment]
Any ideas why some of the tests fail with the future parser?

**Comment by danzilio** [comment]
It looks like the `epel` module requires both the `operatingsystem` and `osfamily` facts. With the future parser, `undef` is (finally) a distinct data type. [Here](https://github.com/stahnma/puppet-module-epel/blob/master/manifests/init.pp#L50), you'll see the `epel` module is trying to perform a regex match against an `undef` (the nonexistent `operatingsystem` fact) and is failing because regex matches only work against the String data type. If you add the `operatingsystem` fact to the RedHat fact hash, the tests should pass.

I like your approach to the fact hashes. I'm going to refactor my tests to do the same!

Could you squash this into a single commit? I'll merge it in after that!

Thanks for your work on this, it's a great addition to the module!

**Comment by mark0n** [comment]
Bingo - that fixes is :-) I pushed a squashed version.

**Comment by danzilio** [comment]
Beautiful! I'll push a new version to the Forge sometime this week!
